File: generics_sorting.cs
using System; 
using System.Collections.Generic; 

using  System.Collections; 


public class GenericSorting{

    public class Person : IComparable<Person>{
        public String name;
        public int age;

        public Person(String name, int age){
            this.name = name;
            this.age =  age;
        }

        public int CompareTo(Person other){ // implement the interface IComaparable function
            if(other == null) // if other is null, this object is placed on sorting algorithm
                return 1;
            
            if(this.age == other.age){
                return this.name.CompareTo(other.name);
            }
            else return this.age.CompareTo(other.age);
            
        }
    }

    public static void createPeople(List<Person> people){
        people.Add(new Person("Hal", 20));
        people.Add(new Person("Sussan", 31));
        people.Add(new Person("Dwight", 19));
        people.Add(new Person("Kassandra", 21));
        people.Add(new Person("Lawrence", 25));
        people.Add(new Person("Cory", 27));
        people.Add(new Person("Mac", 19));
        people.Add(new Person("Romana", 27));
        people.Add(new Person("Doretha", 32));
        people.Add(new Person("Danna", 20));
        people.Add( new Person("Zara", 23));
        people.Add( new Person("Rosalyn", 26));
        people.Add( new Person("Risa", 24));
        people.Add( new Person("Benny", 28));
        people.Add( new Person("Juan", 33));
        people.Add( new Person("Natalie", 25));
        people.Add( new Person("Cindy", 22));
    }


    public static void Main(){
        float[] floats = new float[12] {645.32f, 37.40f, 76.30f, 5.40f, -34.23f, 1.11f, -34.94f, 23.37f, 635.46f, -876.22f, 467.73f, 62.26f};
        List<float> floatList = new List<float>(floats.Length);
        // foreach(float f in floats){
        //    floatList.Add(f);
        // }
        floatList.Sort();
        // foreach(float f in floatList){
        //     Console.WriteLine(f);
        // }
        List<Person> people = new List<Person>();
        createPeople(people);
        foreach(Person p in people){
            Console.WriteLine("{0}, {1}", p.name, p.age);
        }
        Console.WriteLine("");
        people.Sort();
        foreach(Person p in people){
            Console.WriteLine("{0}, {1}", p.name, p.age);
        }
    }
}

File: generics_sorting.hs

import Data.List

--driver
main :: IO()
main = do 
    let floats = [645.32, 37.40, 76.30, 5.40, -34.23, 1.11, -34.94, 23.37, 635.46, -876.22, 467.73, 62.26]
    print $ mySort float_comp floats
    let people = createPeople
    print $ compare (personName (people!!1)) (personName (people!!0))
    print $ mySort people_comp people


                                --Person Class--
data Person = Person String Int deriving Show
    --data members--
personAge :: Person -> Int 
personName :: Person -> String
personAge (Person _ a) = a
personName (Person n _) = n

    --implementation of Eq --
instance Eq Person where 
    (Person n1 a1) == (Person n2 a2) = n1 == n2 && a1 == a2

--Comparator function for Person objects
people_comp :: Person->Person->Ordering
people_comp a b     | personAge a == personAge b = compare (personName a) (personName b)
                    | otherwise = compare (personAge a) (personAge b)
                                 --Person Class--


--Custom sorting generic function
--takes a comparator function and a list of elements, returns a list of elements
mySort :: (a->a->Ordering)->[a] -> [a]
mySort _ [] = []
mySort compFunc xs  = sortBy compFunc xs


-- function to create a list of Person objects
createPeople :: [Person] 
createPeople = [Person "Mac "19, Person "Hal" 20, Person "Sussan" 31, Person "Dwight" 19, Person "Kassandra" 21, Person "Lawrence" 25, Person "Cindy" 22, 
          Person "Cory" 27,  Person "Romana" 27, Person "Doretha" 32, Person "Danna" 20, Person "Zara" 23, Person "Rosalyn" 26, 
           Person "Risa" 24, Person "Benny" 28, Person "Juan" 33, Person "Natalie" 25]

--Comparator function for Floats 
float_comp :: Float->Float->Ordering
float_comp a b = compare a b